lib/src/puppetdb-cli.cc

```diff
@@ -31,49 +33,111 @@ version()
+parse_config(const string& config_content,
+             const string& urls,
+             const string& cacert,
+             const string& cert,
+             const string& key) {
     const json::JsonContainer raw_config(config_content);
     const auto puppetdb_conn = raw_config.includes("puppetdb") ?
-            PuppetDBConn(raw_config.get<json::JsonContainer>("puppetdb")):
-            PuppetDBConn();
+            PuppetDBConn(raw_config.get<json::JsonContainer>("puppetdb"),
+                         urls, cacert, cert, key):
+            PuppetDBConn(urls, cacert, cert, key);
     if (puppetdb_conn.getServerUrl().empty()) {
         throw std::runtime_error { "invalid `server_urls` in configuration" };
     }
     return puppetdb_conn;
 }
 
+vector<string> split(string str, char delimiter) {
+    vector<string> internal;
+    stringstream ss(str);
+    string tok;
+
+    while (getline(ss, tok, delimiter)) {
+        internal.push_back(tok);
+    }
+
+    return internal;
+}
+
+server_urls_t
+parse_server_urls_str(const string& urls) {
+    return split(urls, ',');
+}
+
+
+server_urls_t
+parse_server_urls(const json::JsonContainer& config) {
+    if (config.includes("server_urls")) {
+        const auto urls_type = config.type("server_urls");
+        if (urls_type == json::DataType::Array) {
+            return config.get<server_urls_t>("server_urls");
+        } else if (urls_type == json::DataType::String) {
+            return parse_server_urls_str(config.get<string>("server_urls"));
+        } else {
+            return {};
+        }
+    } else {
+        return {"http://127.0.0.1:8080"};
+    }
+}
+
 PuppetDBConn
-get_puppetdb(const string& config_path) {
+get_puppetdb(const string& config_path,
+             const string& urls,
+             const string& cacert,
+             const string& cert,
+             const string& key) {
     const auto config_content = read_config(config_path);
-    return config_content.empty() ? PuppetDBConn() : parse_config(config_content);
+    return config_content.empty() ? PuppetDBConn(urls, cacert, cert, key)
+            : parse_config(config_content, urls, cacert, cert, key);
 }
 
 PuppetDBConn::PuppetDBConn() :
-        server_urls_ { { "http://127.0.0.1:8080" } },
-        cacert_ { "" },
-        cert_ { "" },
-        key_ { "" } {};
-
-
-PuppetDBConn::PuppetDBConn(const json::JsonContainer& config) :
-        server_urls_ { parseServerUrls(config) },
-        cacert_ { "" },
-        cert_ { "" },
-        key_ { "" } {
-            if (config.includes("cacert")) {
+        server_urls_ { { "http://127.0.0.1:8080" } } ,
+        cacert_ {},
+        cert_ {},
+        key_ {} {};
+
+PuppetDBConn::PuppetDBConn(const string& urls,
+                           const string& cacert,
+                           const string& cert,
+                           const string& key) :
+        server_urls_ { urls.empty() ?
+            vector<string>({ "http://127.0.0.1:8080" }) :
```

Comment:
Huh, bummer. I guess wanting I'm really looking for is the help output showing what the default is, right now it shows no default because its here. Maybe our way of documenting defaults is the wrong pattern or something ... hrm.